### DISCUSSION TITLE
Use 2x BME280 sensors failed

### DISCUSSION THREAD

**Discussion by micheld93** [micropython/micropython#11978] [RP2040 / Pico] [answered]
Hi, 

I need two BME280 sensors which I will connect to my pico. My code is the following: 

```
import machine
import time
import BME280

# Initialize I2C bus 0
i2c_0 = machine.I2C(0, scl=machine.Pin(9), sda=machine.Pin(10), freq=100000)
 
# Initialize I2C bus 1
i2c_1 = machine.I2C(1, scl=machine.Pin(11), sda=machine.Pin(12), freq=100000)

# Create BME280 sensor instances on bus 0
bme1 = BME280.BME280(i2c=i2c_0, address=0x76)

# Create BME280 sensor instances on bus 1
bme2 = BME280.BME280(i2c=i2c_1, address=0x77)

# Read and print temperature and humidity from the first sensor
temperature_1 = bme1.temperature
humidity_1 = bme1.humidity
print("Sensor 1 - Temperature: {:.2f} °C, Humidity: {:.2f} %".format(temperature_1, humidity_1))

# Read and print temperature and humidity from the second sensor
temperature_2 = bme2.temperature
humidity_2 = bme2.humidity
print("Sensor 2 - Temperature: {:.2f} °C, Humidity: {:.2f} %".format(temperature_2, humidity_2))

```

I get this error: 

```
MPY: soft reboot
Traceback (most recent call last):
  File "<stdin>", line 12, in <module>
  File "bme280.py", line 74, in __init__
OSError: [Errno 5] EIO
```

The i2c scanner can detect the sensors when I connect them after another: Same address for both (118 or 0x76). Is something wrong with the code or is it a i2c bus problem?

**Comment by dhalbert** [comment]
> The i2c scanner can detect the sensors when I connect them after another: Same address for both (118 or 0x76). 

If they are both at 0x76, then the initialization below will not find the second one, because it uses address 0x77.

```py
bme2 = BME280.BME280(i2c=i2c_1, address=0x77)
```

On the sensor, if the SDO pin is grounded, it will use address 0x76. If the SDO pin is connected to +V, it will use address 0x77. Don't let it float: then the address is undefined. If you set the address of one sensor to 0x76 and the other to 0x77, you can then connect them to the same I2C bus, and you won't need two I2C buses.

**Reply by micheld93** [reply]
Thanks, I will try it. My sensor break out board has 6 pins and one of them is the SDO. I haven't checked if the SDO is grounded or not but since the pin is reachable I would hope I can get 2 different addresses as you described. 

Just to have the chance to learn a bit more: Is it possible to use the same address when every sensor is connected to its own bus? I would think yes, but I'm not sure. What do I need to change in the code to use the 2 i2c bus on my Pico?

**Comment by dhalbert** [comment]
Could you post a purchase or product link to your sensor board?

The code you posted above is already using two different buses, `i2c_0` and `i2c_1`; there is nothing to change. So you can use two I2C devices with the same address, one on each bus. The alternative is to use a single bus, and connect both devices to the same SCL and SDA pins. Then they must have different addresses.

**Comment by micheld93** [comment]
Sure, I have ordered them here: https://www.berrybase.de/gy-bme280-breakout-board-3in1-sensor-fuer-temperatur-luftfeuchtigkeit-und-luftdruck

Ok, well, then I don't understand why it is not compiling correctly. This should work without correct addresses, right? If not, it could be an address issue since the sensors use the same one. I have connected VCC, GND, SDA and SCL.

**Reply by dhalbert** [reply]
I see two problems:

### Wrong pin numbers
You are using the wrong pin numbers, it appears. You are using the (incomplete and mostly implied) silkscreen pin numbers, but these do not correspond to the GPIO numbers. You want to choose a pair of pins that both I2C0 or both I2C1. See this partial diagram:
![image](https://github.com/micropython/micropython/assets/2847802/769f3095-ac41-4045-8b07-ba69e56f5fd3)
This code is using the wrong numbers. I assume you are using the silkscreen numbers:
```py
# Initialize I2C bus 0
i2c_0 = machine.I2C(0, scl=machine.Pin(9), sda=machine.Pin(10), freq=100000)
 
# Initialize I2C bus 1
i2c_1 = machine.I2C(1, scl=machine.Pin(11), sda=machine.Pin(12), freq=100000)
```
Silkscreen pin 9 is GP6 (`machine.pin(6)`), and silkscreen pin 10 is GP7, and similarly for the other pair. Refer to the pin names on the bottom of the board, not the top. (It would be much nicer if the pin names were on the top of the board.) Also, I will swap the bus names to correspond to the Pico names for the two available I2C periperhals. So it should be:
```py
# Initialize I2C bus 1
i2c_1 = machine.I2C(0, scl=machine.Pin(7), sda=machine.Pin(6), freq=100000)
 
# Initialize I2C bus 0
i2c_0 = machine.I2C(1, scl=machine.Pin(9), sda=machine.Pin(8), freq=100000)
```
### Wrong address for one sensor
Now, you need to use the same address for both sensors, because they are both using the default address, 0x76. Again, I have reversed the bus names to correspond to the names shown in the diagram above.
```py
# Create BME280 sensor instances on bus 1
bme1 = BME280.BME280(i2c=i2c_1, address=0x76)

# Create BME280 sensor instances on bus 0
bme2 = BME280.BME280(i2c=i2c_0, address=0x76)   # not 0x77
```

Alternatively, you could put both sensors on the same I2C bus, and drop the other bus. In that case, you have to tie the SDO pin on one sensor only to +V.

Disclaimer: I am a CircuitPython core developer with I2C experience, hence my reply. I don't use MicroPython regularly.

**Reply by micheld93** [reply]
Thanks for helping and sorry for the delay in answering. I tried your code but it also doesn't work. I get the message: 

`MPY: soft reboot
Traceback (most recent call last):
  File "<stdin>", line 6, in <module>
ValueError: bad SCL pin`

I think the code has mixed bus numbers. The pins GP7 and GP6 are bus 1 not 0 but after "machine.I2C(" the "0" takes bus 0 not 1. Same for bus 0 init.  

```
# Initialize I2C bus 1 
i2c_1 = machine.I2C(0, scl=machine.Pin(7), sda=machine.Pin(6), freq=100000)
 
# Initialize I2C bus 0
i2c_0 = machine.I2C(1, scl=machine.Pin(9), sda=machine.Pin(8), freq=100000)
```


If I'm right and change it to
```
# Initialize I2C bus 1 
i2c_1 = machine.I2C(1, scl=machine.Pin(7), sda=machine.Pin(6), freq=100000)
 
# Initialize I2C bus 0
i2c_0 = machine.I2C(0, scl=machine.Pin(9), sda=machine.Pin(8), freq=100000)

time.sleep_ms(100)
```

The delay is from the link above. It doesn't affect anything. 

```
I get the same error: 
MPY: soft reboot
Traceback (most recent call last):
  File "<stdin>", line 14, in <module>
  File "BME280.py", line 74, in __init__
OSError: [Errno 5] EIO
>>> 
```


When I try to use just one bus and take SDO to VCC for one sensor, I get the same error with EIO. The code is then: 

```
import machine
import time
import BME280

# Initialize I2C bus 1 
# i2c_1 = machine.I2C(1, scl=machine.Pin(7), sda=machine.Pin(6), freq=100000)
 
# Initialize I2C bus 0
i2c_0 = machine.I2C(0, scl=machine.Pin(9), sda=machine.Pin(8), freq=100000)

time.sleep_ms(100)

# Create BME280 sensor instances on bus 0
bme1 = BME280.BME280(i2c=i2c_0, address=0x76)

# Create BME280 sensor instances on bus 1
bme2 = BME280.BME280(i2c=i2c_0, address=0x77)

# Read and print temperature and humidity from the first sensor
temperature_1 = bme1.temperature
humidity_1 = bme1.humidity
print("Sensor 1 - Temperature: {:.2f} °C, Humidity: {:.2f} %".format(temperature_1, humidity_1))

# Read and print temperature and humidity from the second sensor
temperature_2 = bme2.temperature
humidity_2 = bme2.humidity
print("Sensor 2 - Temperature: {:.2f} °C, Humidity: {:.2f} %".format(temperature_2, humidity_2))
```


My next step was to use only one sensor to test the general principle: 

```
from machine import Pin, I2C        #importing relevant modules & classes
from time import sleep

import bme280       #importing BME280 library


i2c_0 = machine.I2C(0, scl=machine.Pin(9), sda=machine.Pin(8), freq=100000)    #initializing the I2C method 

while True:
    bme = bme280.BME280(i2c=i2c)          #BME280 object created
    print(bme.values)
    sleep(2)           #delay of 2s
```
    
But even then I get the EIO error message. And now I'm confused. I added a picture of the wiring for the version with same bus but different addresses. The sensor has the following pins when we look at it from above and the O indicate the mounting points: 

VCC                O
GND
SCL
SDA
CSE
SDO                O



![IMG_1670](https://github.com/micropython/micropython/assets/7882686/4bbfea96-6804-4ad1-aa92-40d6eb6dbb49)

**Reply by micheld93** [reply]
Setup with labeled sensor pins. 
![IMG_1670](https://github.com/micropython/micropython/assets/7882686/e19eb067-d2b9-40a4-82b1-05576b7dced5)

**Comment by bidrohini** [comment]
Here is an old thread about the same error message: https://forum.micropython.org/viewtopic.php?t=4746
It may help you.

**Reply by robert-hh** [reply]
If I count the pins right on the picture, you have connected SDA and SCL to GPIO 11 ans 12. That cannot work. These pins do not even belong to the same I2C device.  At least, a i2c_0.scan() must return both addresses, and only these. EIO is typically an indication of wrong connection.

**Reply by micheld93** [reply]
Yes, it is connected to GPIO 11 and 12. Based on the pinout above they correspond to GP8 and GP9 which are I2C0 SDA and SCL (blue). I think I don't get what you mean with "These pins do not even belong to the same I2C device"

**Reply by robert-hh** [reply, marked as answer]
GPx and GPIOx are just different names for the same thing. And the numbers of the GPIO constructor refer to GPIO/GP numbers. The usual confusion with RPi Pico boards is between GPIO/GP numbers and board pin numbers.
The former start at 0 up to 31, the latter start at 1 up to 40.  So GPIO/GP 8 is at board pin #11, GPIO/GP9 is at board pin #12.

**Reply by micheld93** [reply]
Ok, true. I changed the numbers to the right GPIOs. I use pins GP8/#11 and GP9/#12 for sensor 1 and GP14/#14 and GP15/#15 for sensor 2. 

The code is: 

```
import machine
import time
import BME280

# Initialize I2C bus 0
i2c_0 = machine.I2C(0, scl=machine.Pin(9), sda=machine.Pin(8), freq=100000)

# Initialize I2C bus 1 
i2c_1 = machine.I2C(1, scl=machine.Pin(11), sda=machine.Pin(10), freq=100000)
 

time.sleep_ms(100)

# Create BME280 sensor instances on bus 0
bme1 = BME280.BME280(i2c=i2c_0, address=0x76)

# Create BME280 sensor instances on bus 1
bme2 = BME280.BME280(i2c=i2c_1, address=0x76)

# Read and print temperature and humidity from the first sensor
temperature_1 = bme1.temperature
humidity_1 = bme1.humidity
print("Sensor 1 - Temperature: {:.2f} °C, Humidity: {:.2f} %".format(temperature_1, humidity_1))

# Read and print temperature and humidity from the second sensor
temperature_2 = bme2.temperature
humidity_2 = bme2.humidity
print("Sensor 2 - Temperature: {:.2f} °C, Humidity: {:.2f} %".format(temperature_2, humidity_2))
```
I use this lib https://pypi.org/project/micropython-bme280/ and changed the I2C address to the default value 0x76. 

I checked the wiring and think that it should be right now. Is it true?:
![IMG_1671 2](https://github.com/micropython/micropython/assets/7882686/226cc0a4-498d-4309-b052-6536c181922f)

The error message is still the same but now with different line number. It's now 18: 
```
MPY: soft reboot
Traceback (most recent call last):
  File "<stdin>", line 18, in <module>
  File "BME280.py", line 74, in __init__
OSError: [Errno 5] EIO
```

**Reply by micheld93** [reply]
Ok, for one sensor it is working now but as soon as I add the code for the second sensor it doesn't work anymore.
For whatever reason I needed to use "`temperature1 = bme1.values[0]`" instead of "`temperature_1 = bme1.temperature`". All examples I found on the internet used the "`bme1.temperature`"
I still have the same error OSError: [Errno 5] EIO for the second bus. Even when I init I2C bus 1 alone it does not work. 

```
import machine
import time
import BME280

# Initialize I2C bus 0
i2c_0 = machine.I2C(0, scl=machine.Pin(9), sda=machine.Pin(8), freq=400000)

time.sleep_ms(100)

# Initialize I2C bus 1 
i2c_1 = machine.I2C(1, scl=machine.Pin(15), sda=machine.Pin(14), freq=400000)
 

time.sleep_ms(100)

while True:
    # Create BME280 sensor instances on bus 0
    bme1 = BME280.BME280(i2c=i2c_0, address=0x76)
    temperature1 = bme1.values[0]
    hum1 = bme1.values[1]
    pre1 = bme1.values[2]
    print('Temperatur ist: ', temperature1)
    # Create BME280 sensor instances on bus 0
    time.sleep_ms(100)
    bme2 = BME280.BME280(i2c=i2c_1, address=0x76)
    temperature2 = bme2.values[0]
    hum2 = bme2.values[1]
    pre2 = bme2.values[2]
    print('Temperatur ist: ', temperature2)
```

**Reply by micheld93** [reply]
Well, sometimes a restart of Thonny is the same as doing magic. It is measuring now correctly on both i2c bus numbers.

**Reply by robert-hh** [reply]
Glad that you found your way through the jungle. Thonny tries to be convenient an dsupportive, but sometimes it turns into the opposite.

**Reply by micheld93** [reply]
Me too. Thanks for helping. It was a jungle indeed since I was blind for everything.
